[EDR Workflows][Serverless] Fix Host Isolation Exceptions PLIs

szwarckonrad opened the pull request:
This PR establishes appropriate privileges across Endpoint tiers. In Endpoint Complete, users will be able to create, update, and delete Host Isolation Exceptions. Upon downgrading to Endpoint Essentials, users should still be able to view and remove items from the Blocklist but will not have the ability to edit or create new entries.

Please see linked issue for more details on current behaviour.

Expected privileges:
Endpoint Essentials - User can read and delete HIE
Endpoint Complete - User can read, delete and write HIE

https://github.com/user-attachments/assets/69d14a51-6004-45b7-9c78-62066441f4ab

szwarckonrad committed:
proper pli for host isolation exceptions

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!

szwarckonrad committed:
Merge branch 'main' into fix-host-isolation-exceptions-pli

szwarckonrad committed:
New ProductFeatureKey

szwarckonrad committed:
cleanup

szwarckonrad committed:
Merge branch 'main' into fix-host-isolation-exceptions-pli

szwarckonrad committed:
Merge branch 'main' into fix-host-isolation-exceptions-pli

elasticmachine commented:
Pinging @elastic/security-defend-workflows (Team:Defend Workflows)

szwarckonrad commented:
@gergoabraham 
https://github.com/szwarckonrad/kibana/blob/fix-host-isolation-exceptions-pli/x-pack/plugins/security_solution/server/endpoint/services/artifacts/manifest_manager/manifest_manager.ts#L161-L164
Could you help me out with understanding whether this part needs to be updated? 

My understanding is, that with my changes, this `if` can be simplified to check only for `this.productFeaturesService.isEnabled(ProductFeatureKey.endpointArtifactManagement))` - I moved `SecuritySubFeatureId.hostIsolationExceptions` to `ProductFeatureSecurityKey.endpointArtifactManagement`. However, `SecuritySubFeatureId.hostIsolationExceptions` itself is only targeting Essentials with Read and Delete privileges. I've added `ProductFeatureSecurityKey.endpointHostIsolationExceptions` that targets Complete and extends `SecuritySubFeatureId.hostIsolationExceptions` with write capabilities. Since your mentioned function gets cached artifacts, I believe `SecuritySubFeatureId.hostIsolationExceptions` with it's read and delete only privileges is enough for this check.

semd commented on x-pack/packages/security-solution/features/src/security/product_feature_config.ts:
```diff
@@ -79,10 +79,17 @@ export const securityDefaultProductFeaturesConfig: DefaultSecurityProductFeature
 
   [ProductFeatureSecurityKey.endpointArtifactManagement]: {
     subFeatureIds: [
+      SecuritySubFeatureId.hostIsolationExceptions,
```
NIT: Maybe it would be worth renaming this ID to `SecuritySubFeatureId.hostIsolationExceptionsBasic` since it's included in the `endpointArtifactManagement` PLI (essentials) and only gives _read_ and _delete_ privileges.

semd commented on x-pack/packages/security-solution/features/src/security/product_feature_config.ts:
```diff
@@ -79,10 +79,17 @@ export const securityDefaultProductFeaturesConfig: DefaultSecurityProductFeature
+
+  // Endpoint Complete Tier:
+  // Allows access to create/update HIEs
+  [ProductFeatureSecurityKey.endpointHostIsolationExceptions]: {
+    subFeatureIds: [SecuritySubFeatureId.hostIsolationExceptions],
```
NIT: It's not strictly necessary to add the `subFeatureIds: [SecuritySubFeatureId.hostIsolationExceptions]` here since it's already included in _essentials_ tier by the `ProductFeatureSecurityKey.endpointArtifactManagement` config. 
Nevertheless, it does not have any bad implications either, _subFeatureIds_ are deduplicated internally. As you prefer.

semd commented on x-pack/plugins/security_solution_serverless/common/pli/pli_config.ts:
```diff
@@ -38,10 +38,10 @@ export const PLI_PRODUCT_FEATURES: PliProductFeatures = {
       ProductFeatureKey.endpointResponseActions,
       ProductFeatureKey.osqueryAutomatedResponseActions,
       ProductFeatureKey.endpointAgentTamperProtection,
-      ProductFeatureKey.endpointExceptions,
```
thanks 🙇

dasansol92 replied:
@ashokaditya , is Endpoint Exceptions an Essentials feature?
cc: @caitlinbetz

ashokaditya replied:
Yes endpoint exceptions is available to Endpoint Essentials/Complete

semd reviewed (approved):
Added a couple of NITs. The config changes LGTM!

szwarckonrad commented:
Thank you @semd! Gonna rename to a `Basic` suffixed name. As for the second nit - I would leave it since I believe it makes this part easier to understand and follow. 

Could you take a look at this comment though? - https://github.com/elastic/kibana/pull/191954#issuecomment-2328468785

ashokaditya commented:
> @gergoabraham @semd (since both of you edited this part ;) )
> 
> https://github.com/szwarckonrad/kibana/blob/fix-host-isolation-exceptions-pli/x-pack/plugins/security_solution/server/endpoint/services/artifacts/manifest_manager/manifest_manager.ts#L161-L164
> 
> Could you help me out with understanding whether this part needs to be updated?
> 
> My understanding is, that with my changes, this `if` can be simplified to check only for
> 
> `this.productFeaturesService.isEnabled(ProductFeatureKey.endpointArtifactManagement))`
> 
> I moved `SecuritySubFeatureId.hostIsolationExceptions` to `ProductFeatureSecurityKey.endpointArtifactManagement`.
> 
> However, `SecuritySubFeatureId.hostIsolationExceptions` itself is only targeting Essentials with Read and Delete privileges. I've added `ProductFeatureSecurityKey.endpointHostIsolationExceptions` that targets Complete and extends `SecuritySubFeatureId.hostIsolationExceptions` with write capabilities. Since your mentioned function gets cached artifacts, I believe `SecuritySubFeatureId.hostIsolationExceptions` with it's read and delete only privileges is enough for this check.

That is correct.

semd commented:
@szwarckonrad 
> Could you help me out with understanding whether this part needs to be updated?

I am not familiar with this code, I only refactored the service name, the original author of this conditional is @dasansol92. 
It would need to be good to clarify why we are doing 
`listId === ENDPOINT_ARTIFACT_LISTS.hostIsolationExceptions.id` to know if we need to check the _essentials_ (read + delete) or the _complete_ (write) `ProductFeatureKey`.
I assume we only allow retrieving the exceptions for _essentials_ tier when it's a "host isolation", for the rest (I guess there are other types) we don't even allow reading them. David, can you confirm that?

szwarckonrad committed:
Merge branch 'refs/heads/main' into fix-host-isolation-exceptions-pli

szwarckonrad committed:
cr

szwarckonrad commented on x-pack/plugins/security_solution/server/endpoint/services/artifacts/manifest_manager/manifest_manager.ts:
```diff
@@ -157,12 +157,7 @@ export class ManifestManager {
-        (listId !== ENDPOINT_ARTIFACT_LISTS.hostIsolationExceptions.id &&
-          this.productFeaturesService.isEnabled(ProductFeatureKey.endpointArtifactManagement))
-      ) {
+      if (this.productFeaturesService.isEnabled(ProductFeatureKey.endpointArtifactManagement)) {
```
@dasansol92 Can you confirm that this change is needed and a valid one?

dasansol92 replied:
Not, I don't think we should change this. If there are host isolation exceptions in place but there is a downgrade scenario, those shouldn't be taken into account when generating artifacts. That was the intention of that condition.

szwarckonrad replied:
Reverted

semd replied:
Ok, but the `ProductFeatureKey`s that grant HIE privileges have changed. Shouldn't the condition now be:
```
if (
        (listId === ENDPOINT_ARTIFACT_LISTS.hostIsolationExceptions.id &&
          this.productFeaturesService.isEnabled(ProductFeatureKey.endpointHostIsolationExceptions)) ||
        (listId !== ENDPOINT_ARTIFACT_LISTS.hostIsolationExceptions.id &&
          this.productFeaturesService.isEnabled(ProductFeatureKey.endpointArtifactManagement))
      ) {
```

semd replied:
It would also be a good idea to add a comment explaining this condition to avoid confusion in the future

dasansol92 replied:
I think @semd is right here since the HIE feature has been moved outside of `endpointResponseActions`

szwarckonrad replied:
6e50a73b0337754c312730c8ff984a52109ecf86
Changed to what @semd proposed, I agree that this makes the most sense when the purpose of this check is now clear. Thanks!
cc @dasansol92

semd commented on x-pack/packages/security-solution/features/src/security/kibana_sub_features.ts:
```diff
@@ -125,7 +125,7 @@ const trustedApplicationsSubFeature: SubFeatureConfig = {
   ],
 };
-const hostIsolationExceptionsSubFeature: SubFeatureConfig = {
+const hostIsolationExceptionsSubFeatureBasic: SubFeatureConfig = {
```
```suggestion
const hostIsolationExceptionsBasicSubFeature: SubFeatureConfig = {
```

szwarckonrad replied:
Changed in 701474a6f447d0f5270de971873d5133edece4fb

semd commented on x-pack/packages/security-solution/features/src/product_features_keys.ts:
```diff
@@ -121,7 +126,7 @@ export enum SecuritySubFeatureId {
   endpointExceptions = 'endpointExceptionsSubFeature',
   trustedApplications = 'trustedApplicationsSubFeature',
-  hostIsolationExceptions = 'hostIsolationExceptionsSubFeature',
+  hostIsolationExceptionsBasic = 'hostIsolationExceptionsSubFeature',
```
```suggestion
  hostIsolationExceptionsBasic = 'hostIsolationExceptionsBasicSubFeature',
```

szwarckonrad replied:
Changed in 701474a6f447d0f5270de971873d5133edece4fb

szwarckonrad committed:
cr

szwarckonrad committed:
Merge branch 'refs/heads/main' into fix-host-isolation-exceptions-pli

szwarckonrad committed:
cr

szwarckonrad committed:
cr

semd reviewed (approved):
Everything looks much cleaner now 💯 
Thanks. LGTM!

gergoabraham reviewed (commented):
nice work, looks good!

did some downgrade tests with it with a local serverless instance,
- `endpoint:complete` -> `endpoint:none` looks good, everything protection related is hidden
- `endpoint:complete` -> `endpoint:essentials` also looks good, HIE can only be removed, and then after refreshing the page, the user cannot access it ✅ 
![downgrade](https://github.com/user-attachments/assets/b245b91e-c183-4ea2-b878-ae7e33af5f4a)

❓ one question: is it intended to leave the HIE tab visible on the Policy Details page for Endpoint Essentials? (HIE is hidden from the Assets in this case)
<img width="400" alt="image" src="https://github.com/user-attachments/assets/259dc4ce-38e9-4c1d-a0ce-9bef277224a5">

szwarckonrad commented:
@gergoabraham  
https://github.com/elastic/kibana/issues/191945 :)

gergoabraham reviewed (approved):
ah, so this is a prerequisite, cool 👍
